cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,359 @@
+		// cleanup firewall and service cache
+		fc.svcCache.delete(key)
+		fc.fwCache.delete(key)
+	case err != nil:
+		runtime.HandleError(fmt.Errorf("Unable to retrieve service %v from store: %v", key, err))
+	default:
+		err = fc.processServiceCreateOrUpdate(service, key)
+	}
+
+	return err
+}
+
+// processServiceCreateOrUpdate modifies (or not) the firewall for the incoming service accordingly.
+// Returns an error if processing the service update failed.
+func (fc *Controller) processServiceCreateOrUpdate(service *v1.Service, key string) error {
+	cachedService := fc.svcCache.getOrCreate(key)
+	if cachedService.state != nil && cachedService.state.UID != service.UID {
+		// This happens only when a service is deleted and re-created
+		// in a short period, which is only possible when it doesn't
+		// contain finalizer.
+		if err := fc.processFirewallDelete(key, "", cachedService.state); err != nil {
+			return err
+		}
+	}
+	// Always cache the service, we need the info for firewall modification.
+	cachedService.state = service
+
+	return nil
+}
+
+func (fc *Controller) processFirewallCreateOrUpdate(ctx context.Context, key, workerFirewallName string) error {
+	cachedFirewall := fc.fwCache.getOrCreate(key)
+	// We do not care about other firewalls
+	if cachedFirewall.state.Name != workerFirewallName {
+		return nil
+	}
+
+	// Inbound rules firewall request setup for NodePort range.
+	inboundNodePortRules := &godo.InboundRule{
+		Protocol:  "tcp",
+		PortRange: "30000-32767",
+	}
+	rr := &godo.FirewallRulesRequest{
+		InboundRules: []godo.InboundRule{*inboundNodePortRules},
+	}
+	fr := &godo.FirewallRequest{
+		Name:         workerFirewallName,
+		InboundRules: []godo.InboundRule{*inboundNodePortRules},
+	}
+
+	// Create the firewall if it does not exist, update it if it does.
+	fw, _, _ := fc.firewallService.Get(ctx, cachedFirewall.state.ID)
+	if fw == nil {
+		// Create a new firewall with the worker firewall name and inbound rules for NodePort range.
+		fw, err := fc.create(ctx, workerFirewallName, fr, rr)
+		if err != nil {
+			return fmt.Errorf("Failed to create firewall: %s", err)
+		}
+		cachedFirewall.state = fw
+		return nil
+	} else if fw.Name == workerFirewallName {
+		// Update the existing firewall to include inbound rules for NodePort range.
+		fw, err := fc.update(ctx, fw, rr)
+		if err != nil {
+			return fmt.Errorf("Failed to update firewall: %s", err)
+		}
+		cachedFirewall.state = fw
+		return nil
+	}
+	// Always cache the firewall, we need the info for future firewall modification.
+	cachedFirewall.state = fw
+
+	return nil
+}
+
+func (fc *Controller) processFirewallDelete(key, workerFirewallname string, service *v1.Service) error {
+	cachedFirewall := fc.fwCache.getOrCreate(key)
+	// We do not care about other firewalls
+	if cachedFirewall.state.Name != workerFirewallName {
+		return nil
+	}
+	_, err := fc.firewallService.Delete(context.TODO(), cachedFirewall.state.ID)
+	if err != nil {
+		return err
```

Comment:
This should handle the case where the firewall is already gone (404) for whatever reason. In that case, it's not an error.